README.md
# Description

The __wlp__ cookbook installs and configures WebSphere Application Server Liberty Profile.
It provides recipes, resources, and libraries for creating, managing, and configuring Liberty profile server instances.

## Basic configuration

The __wlp__ cookbook can install Liberty profile from jar archives or a zip file. The installation method is configured via the `node[:wlp][:install_method]` attribute. 

### jar installation

When jar archives installation method is used the `node[:wlp][:archive][:base_url]` attribute must be set to specify the base location from which the Liberty jar archives can be downloaded. Use the `node[:wlp][:archive][:runtime][...]`, `node[:wlp][:archive][:extended][...]`, and `node[:wlp][:archive][:extras][...]` attributes (described below) to adjust jar archive file names, checksums, and other installation options. You must also set the `node[:wlp][:archive][:accept_license]` attribute to `true` if you agree and accept the license terms of the Liberty profile software. The jar archive-based installation will fail if `node[:wlp][:archive][:accept_license]` is not set to `true`. 


By default the cookbook is configured with jar archive installation method. The archive installation options are pre-configured with values based on developer licensed Liberty profile. 

The no-fee developer licensed Liberty profile jar archives can be downloaded from [Liberty download page](https://www.ibmdw.net/wasdev/downloads/websphere-application-server-liberty-profile/) on [WASdev](https://www.ibmdw.net/wasdev) site. The production licensed Liberty profile jar archives can be obtained from [IBM Passport Advantage](http://www-01.ibm.com/software/lotus/passportadvantage/).

### zip installation

When zip installation method is used only the `node[:wlp][:zip][:url]` attribute must be set to specify the location of the zip file. The zip is assumed to be generated by running `./bin/server package` Liberty command with `--include=all` or `--include=minify` option.



# Requirements

## Platform:

* Debian
* Ubuntu
* Centos
* Redhat

## Cookbooks:

* java

# Attributes

* `node[:wlp][:user]` - User name under which the server will be installed and running. Defaults to `wlp`.
* `node[:wlp][:group]` - Group name under which the server will be installed and running. Defaults to `wlp-admin`.
* `node[:wlp][:install_java]` - Use the `java` cookbook to install Java. If Java is installed in a different manner 
set it to `false`. The Java executables must be available on the __PATH__. Defaults to `true`.
* `node[:wlp][:base_dir]` - Base installation directory. Defaults to `/opt/was/liberty`.
* `node[:wlp][:user_dir]` - Set user configuration directory (wlp.user.dir). Set to 'nil' to use default location. Defaults to `nil`.
* `node[:wlp][:install_method]` - Installation method. Set it to 'archive' or 'zip'. Defaults to `archive`.
* `node[:wlp][:archive][:base_url]` - Base URL location for downloading the runtime, extended, and extras Liberty profile archives. 
Must be set when `node[:wlp][:install_method]` is set to `archive`. Defaults to `nil`.
* `node[:wlp][:archive][:runtime][:url]` - URL location of the runtime archive. Defaults to `#{node[:wlp][:archive][:base_url]}/wlp-developers-runtime-8.5.5.1.jar`.
* `node[:wlp][:archive][:runtime][:checksum]` - Checksum value for the runtime archive. Defaults to `df6e4cf78f91745a11372f1b4a8467fea8e7c53ddec48471cf92729deb88d306`.
* `node[:wlp][:archive][:extended][:url]` - URL location of the extended archive. Defaults to `#{node[:wlp][:archive][:base_url]}/wlp-developers-extended-8.5.5.1.jar`.
* `node[:wlp][:archive][:extended][:checksum]` - Checksum value for the extended archive. Defaults to `d511fc4068f4e9c345b8fbfa2423fa00e476c2d771dde884a4b962fc3094149c`.
* `node[:wlp][:archive][:extended][:install]` - Controls whether the extended archive should be downloaded and installed. Defaults to `true`.
* `node[:wlp][:archive][:extras][:url]` - URL location of the extras archive. Defaults to `#{node[:wlp][:archive][:base_url]}/wlp-developers-extras-8.5.5.1.jar`.
* `node[:wlp][:archive][:extras][:checksum]` - Checksum value for the extras archive. Defaults to `c468247e18ffb85b1d691f67839471ccd4390b299eb997151f7b56efc6332f4d`.
* `node[:wlp][:archive][:extras][:install]` - Controls whether the extras archive should be downloaded and installed. Defaults to `false`.
* `node[:wlp][:archive][:extras][:base_dir]` - Base installation directory of the extras archive. Defaults to `#{node[:wlp][:base_dir]}/extras`.
* `node[:wlp][:archive][:accept_license]` - Accept license terms when doing archive-based installation. 
Must be set to `true` otherwise installation will fail. Defaults to `false`.
* `node[:wlp][:zip][:url]` - URL location to a zip file containing Liberty profile installation files. Must be set 
only when `node[:wlp][:install_method]` is set to `zip`. Defaults to `nil`.
* `node[:wlp][:config][:basic]` - Defines basic server configuration when creating server instances using the `wlp_server` resource. Defaults to `{ ... }`.
* `node[:wlp][:servers][:defaultServer]` - Defines `defaultServer` server instance. Used by the `serverconfig` recipe. Defaults to `{ ... }`.

# Recipes

* [wlp::default](#wlpdefault) - Installs WebSphere Application Server Liberty Profile.
* [wlp::serverconfig](#wlpserverconfig) - Creates Liberty profile server instance for each `node[:wlp][:servers][<server_name>]` definition.

## wlp::default

Installs WebSphere Application Server Liberty Profile. Liberty profile can be 
installed using jar file archives or from a zip file based on the `node[:wlp][:install_method]` setting.

## wlp::serverconfig

Creates Liberty profile server instance for each `node[:wlp][:servers][<server_name>]` definition.
The following definition creates a simple `airport` server instance:
```ruby
node[:wlp][:servers][:airport] = {
  "enabled" => true,
  "description" => "Airport Demo App",
  "featureManager" => {
    "feature" => [ "jsp-2.2" ]
  },
  "httpEndpoint" => {
    "id" => "defaultHttpEndpoint",
    "host" => "*",
    "httpPort" => "9080",
    "httpsPort" => "9443"
  }
}
```

# Resources

* [wlp_bootstrap_properties](#wlp_bootstrap_properties) - Adds, sets and removes bootstrap properties for a particular server instance.
* [wlp_config](#wlp_config) - Generates server.xml from a hash expression.
* [wlp_install_feature](#wlp_install_feature) - Installs a feature from an enterprise subsystem archive (ESA) file.
* [wlp_jvm_options](#wlp_jvm_options) - Adds, sets and removes JVM options in installation-wide or instance-specific jvm.options file.
* [wlp_server](#wlp_server) - Provides operations for creating, starting, stopping, and destroying Liberty profile server instances.
* [wlp_server_env](#wlp_server_env) - Adds, sets and removes environment properties in installation-wide or instance-specific server.env file.

## wlp_bootstrap_properties

Adds, sets and removes bootstrap properties for a particular server instance.

### Actions

- set: Set properties in bootstrap.properties file. Other existing properties in the file will not be preserved. Default action.
- add: Adds properties to bootstrap.properties file. Other existing properties in the file will be preserved.
- remove: Removes properties from bootstrap.properties file. Other existing properties in the file will be preserved.

### Attribute Parameters

- server_name: Name of the server instance. Defaults to <code>nil</code>.
- properties: The properties to add, set or remove. Must be specified as a hash when adding or setting and as an array when removing. Defaults to <code>nil</code>.

### Examples
```ruby
wlp_bootstrap_properties "add to bootstrap.properties" do
  server_name "myInstance"
  properties "com.ibm.ws.logging.trace.file.name" => "trace.log"
  action :add
end

wlp_bootstrap_properties "remove from bootstrap.properties" do
  server_name "myInstance"
  properties [ "com.ibm.ws.logging.trace.file.name" ]
  action :remove
end

wlp_bootstrap_properties "set bootstrap.properties" do
  properties "default.http.port" => "9081", "default.https.port" => "9444"
  action :set
end
```

## wlp_config

Generates server.xml from a hash expression.

### Actions

- create: Creates/updates server.xml. Default action.
- create_if_missing: Creates server.xml only if the file does not already exist.

### Attribute Parameters

- file: The server.xml file to create or update. Defaults to <code>nil</code>.
- config: The contents of the server.xml expressed as a hash. Defaults to <code>nil</code>.

### Examples
```ruby
wlp_config "/var/servers/airport/server.xml" do
  config ({
            "description" => "Airport Demo App",
            "featureManager" => {
              "feature" => [ "jsp-2.2" ]
            },
            "httpEndpoint" => {
              "id" => "defaultHttpEndpoint",
              "host" => "*",
              "httpPort" => "9080",
              "httpsPort" => "9443"
            }
  })
end
```

## wlp_install_feature

Installs a feature from an enterprise subsystem archive (ESA) file.

### Actions

- install: Installs a feature using .esa file. Default action.

### Attribute Parameters

- location: Specifies the location of the subsystem archive (ESA file) to install. Can be a file name or a URL. Defaults to <code>nil</code>.
- to: Specifies where to install the feature. The feature can be installed to any configured product extension location, or as a user feature. Defaults to <code>"usr"</code>.
- accept_license: Specifies whether to accept the license terms and conditions of the feature. Defaults to <code>false</code>.

### Examples
```ruby
wlp_install_feature "mongodb" do
  location "http://public.dhe.ibm.com/ibmdl/export/pub/software/websphere/wasdev/downloads/wlp/8.5.5.1/com.ibm.websphere.appserver.mongodb-2.0.esa"
  accept_license true
end
```

## wlp_jvm_options

Adds, sets and removes JVM options in installation-wide or instance-specific jvm.options file.

### Actions

- add: Adds JVM options to jvm.options file. Other existing options in the file will be preserved. Default action.
- remove: Removes JVM options from jvm.options file. Other existing options in the file will be preserved.
- set: Sets JVM options in jvm.options file. Other existing options will not be preserved.

### Attribute Parameters

- server_name: If specified, the jvm.options file in the specified server instance is updated. Otherwise, the installation-wide jvm.options file is updated. Defaults to <code>nil</code>.
- options: The JVM options to add, set or remove. Defaults to <code>nil</code>.

### Examples
```ruby
wlp_jvm_options "add to instance-specific jvm.options" do
  server_name "myInstance"
  options [ "-Djava.net.ipv4=true" ]
  action :add
end

wlp_jvm_options "remove from instance-specific jvm.options" do
  server_name "myInstance"
  options [ "-Djava.net.ipv4=true" ]
  action :remove
end

wlp_jvm_options "add to installation-wide jvm.options" do
  options [ "-Xmx1024m" ]
  action :add
end

wlp_jvm_options "remove from installation-wide jvm.options" do
  options [ "-Xmx1024m" ]
  action :remove
end
```

## wlp_server

Provides operations for creating, starting, stopping, and destroying Liberty profile server instances.

### Actions

- start: Creates and starts the server instance (as an OS service). Default action.
- create: Creates/updates server instance.
- create_if_missing: Creates server instance only if the instance does not already exist.
- destroy: Destroys server instance.
- stop: Stops the server instance (via an OS service).

### Attribute Parameters

- server_name: Name of the server instance.
- config: Configuration for the server instance. If not specified, `node[:wlp][:config][:basic]` is used as the initial configuration. Defaults to <code>nil</code>.
- jvmOptions: Instance-specific JVM options. Defaults to <code>[]</code>.
- serverEnv: Instance-specific server environment properties. Defaults to <code>{}</code>.
- bootstrapProperties: Instance-specific bootstrap properties. Defaults to <code>{}</code>.
- clean: Clean all cached information when starting the server instance. Defaults to <code>false</code>.

### Examples
```ruby
wlp_server "myInstance" do 
  config ({
            "featureManager" => {
              "feature" => [ "jsp-2.2", "jaxws-2.1" ]
            },
            "httpEndpoint" => {
              "id" => "defaultHttpEndpoint",
              "host" => "*",
              "httpPort" => "9080",
              "httpsPort" => "9443"
            },
            "application" => {
              "id" => "example",
              "name" => "example",
              "type" => "war",
              "location" => "/apps/example.war"
            }
          })
  jvmOptions [ "-Djava.net.ipv4=true" ]
  serverEnv "JAVA_HOME" => "/usr/lib/j2sdk1.7-ibm/"
  bootstrapProperties "default.http.port" => "9080", "default.https.port" => "9443"
  action :create
end

wlp_server "myInstance" do 
  clean true
  action :start
end

wlp_server "myInstance" do
  action :stop
end

wlp_server "myInstance" do
  action :destroy
end
```

## wlp_server_env

Adds, sets and removes environment properties in installation-wide or instance-specific server.env file.

### Actions

- set: Set environment properties in server.env file. Other existing properties in the file will not be preserved. Default action.
- add: Adds environment properties in server.env file. Other existing properties in the file will be preserved.
- remove: Removes environment properties in server.env file. Other existing properties in the file will be preserved.

### Attribute Parameters

- server_name: If specified, the server.env file in the specified server instance is updated. Otherwise, the installation-wide server.env file is updated. Defaults to <code>nil</code>.
- properties: The properties to add, set or remove. Must be specified as a hash when adding or setting and as an array when removing. Defaults to <code>nil</code>.

### Examples
```ruby
wlp_server_env "add to instance-specific server.env" do
  server_name "myInstance"
  properties "JAVA_HOME" => "/usr/lib/j2sdk1.7-ibm/"
  action :add
end

wlp_server_env "remove from instance-specific server.env" do
  server_name "myInstance"
  properties [ "JAVA_HOME" ]
  action :remove
end

wlp_server_env "set installation-wide server.env" do
  properties "WLP_USER_DIR" => "/var/wlp"
  action :set
end

wlp_server_env "remove from installation-wide server.env" do
  properties [ "WLP_USER_DIR" ]
  action :remove
end
```

# Support

Use the [issue tracker][] for reporting any bugs or enhancements. For any questions please use the [WASdev forum](https://www.ibm.com/developerworks/community/forums/html/forum?id=11111111-0000-0000-0000-000000002666).

[issue tracker]: https://github.com/WASdev/ci.chef.wlp/issues

The cookbook is maintained by IBM.

# Notice

© Copyright IBM Corporation 2013.

# License

```text
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```


CHANGELOG.md
# Cookbook Changelog

This file is used to list changes made in each version of the __wlp__ cookbook.

## 0.2.1:

### Improvements

- **[Issue #16](https://github.com/WASdev/ci.chef.wlp/issues/16)** - IBM Java & Java cookbook

## 0.2.0:

### Improvements

- **[Issue #15](https://github.com/WASdev/ci.chef.wlp/issues/15)** - Add support for setting bootstrap properties
- **[Issue #14](https://github.com/WASdev/ci.chef.wlp/issues/14)** - Make zip\_install & archive\_install recipes private
- **[Issue #13](https://github.com/WASdev/ci.chef.wlp/issues/13)** - Updates for Liberty Profile 8.5.5.1
- **[Issue #12](https://github.com/WASdev/ci.chef.wlp/issues/12)** - Add resource for installing Liberty features (esa files)
- **[Issue #11](https://github.com/WASdev/ci.chef.wlp/issues/11)** - Add support for encoding passwords

## 0.1.0:

* Initial release.



resources/server.rb
# Cookbook Name:: wlp
# Attributes:: default
#
# (C) Copyright IBM Corporation 2013.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

=begin
#<
Provides operations for creating, starting, stopping, and destroying Liberty profile server instances.

@action create  Creates/updates server instance.
@action create_if_missing  Creates server instance only if the instance does not already exist.
@action destroy Destroys server instance.
@action start   Creates and starts the server instance (as an OS service). 
@action stop    Stops the server instance (via an OS service).

@section Examples
```ruby
wlp_server "myInstance" do 
  config ({
            "featureManager" => {
              "feature" => [ "jsp-2.2", "jaxws-2.1" ]
            },
            "httpEndpoint" => {
              "id" => "defaultHttpEndpoint",
              "host" => "*",
              "httpPort" => "9080",
              "httpsPort" => "9443"
            },
            "application" => {
              "id" => "example",
              "name" => "example",
              "type" => "war",
              "location" => "/apps/example.war"
            }
          })
  jvmOptions [ "-Djava.net.ipv4=true" ]
  serverEnv "JAVA_HOME" => "/usr/lib/j2sdk1.7-ibm/"
  bootstrapProperties "default.http.port" => "9080", "default.https.port" => "9443"
  action :create
end

wlp_server "myInstance" do 
  clean true
  action :start
end

wlp_server "myInstance" do
  action :stop
end

wlp_server "myInstance" do
  action :destroy
end
```
#>
=end
actions :start, :stop, :create, :create_if_missing, :destroy

#<> @attribute server_name Name of the server instance.
attribute :server_name, :kind_of => String, :name_attribute => true

#<> @attribute config Configuration for the server instance. If not specified, `node[:wlp][:config][:basic]` is used as the initial configuration.
attribute :config, :kind_of => Hash, :default => nil

#<> @attribute jvmOptions Instance-specific JVM options. 
attribute :jvmOptions, :kind_of => Array, :default => []

#<> @attribute serverEnv Instance-specific server environment properties. 
attribute :serverEnv, :kind_of => Hash, :default => {}

#<> @attribute bootstrapProperties Instance-specific bootstrap properties. 
attribute :bootstrapProperties, :kind_of => Hash, :default => {}

#<> @attribute clean Clean all cached information when starting the server instance.
attribute :clean, :kind_of => [TrueClass, FalseClass], :default => false

default_action :start

